service/history/historyCache.go

```diff
@@ -96,12 +95,45 @@ func (c *historyCache) getOrCreateWorkflowExecutionWithTimeout(ctx context.Conte
 		})
 
 		if err != nil {
-			return nil, nil, err
+			return err
 		}
 
 		execution.RunId = common.StringPtr(response.RunID)
 	} else if uuid.Parse(execution.GetRunId()) == nil { // immediately return if invalid runID
-		return nil, nil, &workflow.BadRequestError{Message: "RunID is not valid UUID."}
+		return &workflow.BadRequestError{Message: "RunID is not valid UUID."}
+	}
+	return nil
+}
+
+// For analyzing mutableState, we have to try get workflowExecutionContext from cache and also load from database
+func (c *historyCache) getAndCreateWorkflowExecutionWithTimeout(ctx context.Context, domainID string,
+	execution workflow.WorkflowExecution) (*workflowExecutionContext, *workflowExecutionContext, releaseWorkflowExecutionFunc, bool, error) {
+	if err := c.validateWorkflowExecutionInfo(domainID, &execution); err != nil {
+		return nil, nil, nil, false, err
+	}
+
+	key := execution.GetRunId()
+	contextFromCache, cacheHit := c.Get(key).(*workflowExecutionContext)
+	releaseFunc := func(error) {}
+	// If cache hit, we need to lock the cache to prevent race condition
+	if cacheHit {
```

Comment:
this feels complicated, it would be simpler if we just return from whatever in the cache, and provide another API to force unload the cache, which is also needed as an admin command.

Reply:
This s about requirement of the task. I will talk to you in uChat